core/src/main/java/com/netflix/iceberg/ReplaceFiles.java

```diff
@@ -16,21 +16,113 @@
+    private final Set<DataFile> filesToAdd = new HashSet<>();
+    private final Set<CharSequence> pathsToDelete = new HashSet<>();
+    private final List<String> newManifests = new ArrayList<>();
+    private List<String> cachedManifests = new ArrayList<>();
+    private final AtomicInteger manifestCount = new AtomicInteger(0);
+    private boolean hasChanges = false;
+
+    public ReplaceFiles(TableOperations ops) {
+        super(ops);
+        this.ops = ops;
+    }
+
+    /**
+     * @param base TableMetadata of the base snapshot
+     * @return list of manifests that "may" get committed if commit is called on this instance.
+     */
+    @Override
+    protected List<String> apply(TableMetadata base) {
+        ValidationException.check(!(this.pathsToDelete.isEmpty() || this.filesToAdd.isEmpty()), "Must provide some files to add and delete.");
+        if (this.hasChanges) {
+            final Snapshot snapshot = base.currentSnapshot();
+            ValidationException.check(snapshot != null, "No snapshots are committed so no datafiles can be replaced.");
+
+            final List<String> currentManifests = snapshot.manifests();
+            final List<CharSequence> deletedPaths = Collections.synchronizedList(new ArrayList<>());
+            final List<String> addedManifests = Collections.synchronizedList(new ArrayList<>());
+
+            currentManifests.parallelStream().forEach(manifest -> {
+                final OutputFile manifestPath = manifestPath(manifestCount.getAndIncrement());
+                try (ManifestReader manifestReader = ManifestReader.read(ops.newInputFile(manifest))) {
+                    try (ManifestWriter writer = new ManifestWriter(manifestReader.spec(), manifestPath, snapshotId())) {
+                        for (ManifestEntry manifestEntry : manifestReader.entries()) {
+                            if (this.pathsToDelete.contains(manifestEntry.file().path()))
+                                if (manifestEntry.status() == ManifestEntry.Status.DELETED) {
+                                    final String msg = String.format("file %s is already marked as deleted in manifest File %s", manifestEntry.file().path(), manifest);
+                                    throw new CommitFailedException(msg);
+                                } else {
+                                    writer.delete(manifestEntry);
+                                    deletedPaths.add(manifestEntry.file().path());
+                                }
+                            else if (manifestEntry.status() != ManifestEntry.Status.DELETED) {
+                                writer.addExisting(manifestEntry);
+                            }
+                        }
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException("Could not read manifestFile " + manifest);
+                } finally {
+                    addedManifests.add(manifestPath.location());
```

Comment:
This always adds the new manifest, but if there were no files deleted from the manifest, I think it should reuse the original manifest. This is what happens in [`StreamingDelete`](https://github.com/Netflix/iceberg/blob/master/core/src/main/java/com/netflix/iceberg/StreamingDelete.java#L153) when there are no files deleted. (Eventually, we might avoid writing and deleting the files if the contents can be held in memory.)

Reply:
Done.